src/java.base/share/classes/java/util/concurrent/Exchanger.java

```diff
@@ -366,47 +365,39 @@ else if ((nb = (b + 1) & MMASK) < alen) {
                     }
                 }
             }
-            else {                               // try offer
-                if (offered == null)
-                    offered = p.item = item;
-                if (ENTRY.compareAndSet(s, null, p)) {
-                    boolean tryCancel;           // true if interrupted
-                    Thread t = Thread.currentThread();
-                    if (!(tryCancel = t.isInterrupted()) && ncpu > 1 &&
-                        (i != 0 ||               // check for busy VTs
-                         (!ForkJoinWorkerThread.hasKnownQueuedWork()))) {
-                        for (int j = SPINS; j > 0; --j) {
-                            if ((v = p.match) != null) {
-                                MATCH.set(p, null);
-                                break outer;     // spin wait
-                            }
-                            Thread.onSpinWait();
+            else if (ENTRY.compareAndSet(s, null, p)) { // try offer
+                boolean tryCancel = false;
+                for (long ns = 1L;;) {
+                    if (p.match == null && !tryCancel) {
+                        if ((deadline != 0L &&
+                             (ns = deadline - System.nanoTime()) <= 0L) ||
+                            Thread.currentThread().isInterrupted())
+                            tryCancel = true;    // cancel unless match
+                        else if (ncpu > 1 &&
+                                 (i != 0 ||      // check for busy VTs
+                                  (!ForkJoinWorkerThread.hasKnownQueuedWork()))) {
+                            for (int j = SPINS; p.match == null && j > 0; --j)
+                                Thread.onSpinWait();
```

Comment:
It might not be worth it, but I wonder if something like this would make a positive impact:

```suggestion
                        if (!(tryCancel = (deadline != 0L &&
                             (ns = deadline - System.nanoTime()) <= 0L) ||
                            Thread.currentThread().isInterrupted()))
                            && ncpu > 1 &&
                             (i != 0 || /* check for busy VTs */ (!ForkJoinWorkerThread.hasKnownQueuedWork()))) {
                            for (int j = SPINS; p.match == null && j > 0; --j)
                                Thread.onSpinWait();
```